src/nvim/lua/vim.lua

```diff
@@ -478,4 +478,105 @@ function vim.defer_fn(fn, timeout)
   return timer
 end
 
+--- Parse a version string into a table.
+---
+--@param ver (string): version string
+--@param strict (optional bool): if true, a semver string is expected;
+--       otherwise, just dotted numbers, possibly with junk at the end
+--@return table with keys major, minor, patch, pre and build if strict,
+--        otherwise with numerical indices
+function vim.parse_version(ver, strict)
+  local err_msg = string.format('invalid version string: "%s"', ver)
+
+  local numbers = {}
+  local cursor, end_number, number
+  repeat
+    _, end_number, number = string.find(ver, '^(%d+)%.?', cursor)
+    if number then
+      cursor = end_number + 1
+      table.insert(numbers, tonumber(number))
+    end
+  until not number
+
+  if #numbers == 0 or (strict and #numbers > 3) then
+    error(err_msg)
+  elseif not strict then
+    return numbers
+  end
+
+  local _, end_pre, pre = string.find(ver, '^%-([0-9A-Za-z.-]+)', cursor)
+  if pre then
+    cursor = end_pre + 1
+  end
+
+  local _, end_build, build = string.find(ver, '^%+([0-9A-Za-z.-]+)', cursor)
+  if build then
+    cursor = end_build + 1
+  end
+
+  if cursor >= #ver then
+    local major, minor, patch = unpack(numbers)
+    return {major=major, minor=minor, patch=patch, pre=pre, build=build}
+  else
+    error(err_msg)
+  end
+end
+
+--- Compare two version strings.
+---
+--@param ver1, ver2 (string): version strings
+--@param strict (optional bool): if true, semver strings are expected;
+--       otherwise, just dotted numbers, possibly with junk at the end
+--@return (number) 1 if ver1 is greater, -1 if it's smaller, 0 otherwise
+function vim.cmp_versions(ver1, ver2, strict)
```

Comment:
The strict mode comments go here as well.

One suggestion might be that there should be a "minimum comparison" key or similar. For example, most of the time I probably just want to check major, minor patch. If they are the same, then I'll call that "equal" -- even if they aren't strictly equal. I'm not sure exactly the best way to handle that though.

Reply:
Re: strict mode, please see https://github.com/neovim/neovim/pull/12437#discussion_r437754800 and let's have that discussion there?

> a "minimum comparison" key or similar

That's a good suggestion, thanks! I'll try and think of a way of incorporating it :)